frontend/src/lib/k8s/apiProxy.ts

```diff
@@ -369,16 +380,30 @@ function simpleApiFactoryWithNamespace(
     scale?: ReturnType<typeof apiScaleFactory>;
     [other: string]: any;
   } = {
-    list: (namespace: string, cb: StreamResultsCb, errCb: StreamErrCb) =>
-      streamResults(url(namespace), cb, errCb),
-    get: (namespace: string, name: string, cb: StreamResultsCb, errCb: StreamErrCb) =>
-      streamResult(url(namespace), name, cb, errCb),
-    post: (body: KubeObjectInterface) => post(url(body.metadata.namespace as string), body),
-    patch: (body: OpPatch[], namespace: string, name: string) =>
-      patch(`${url(namespace)}/${name}?pretty=true`, body),
-    put: (body: KubeObjectInterface) =>
-      put(`${url(body.metadata.namespace as string)}/${body.metadata.name}`, body),
-    delete: (namespace: string, name: string) => remove(`${url(namespace)}/${name}`),
+    list: (
+      namespace: string,
+      cb: StreamResultsCb,
+      errCb: StreamErrCb,
+      queryParams?: QueryParameters
+    ) => streamResults(url(namespace), cb, errCb, queryParams),
+    get: (
+      namespace: string,
+      name: string,
+      cb: StreamResultsCb,
+      errCb: StreamErrCb,
+      queryParams?: QueryParameters
+    ) => streamResult(url(namespace), name, cb, errCb, queryParams),
+    post: (body: KubeObjectInterface, queryParams?: QueryParameters) =>
+      post(url(body.metadata.namespace as string) + asQuery(queryParams), body),
+    patch: (body: OpPatch[], namespace: string, name: string, queryParams?: QueryParameters) =>
+      patch(`${url(namespace)}/${name}` + asQuery({ ...queryParams, ...{ pretty: 'true' } }), body),
```

Comment:
I wonder if we should be using URL and URL.searchParams instead of stringly adding them together. Would be a bit safer.

Reply:
In asQuery, URL searchParams is used. Could you please explain what is stringly adding them and what does safer mean?

Reply:
We had a little conversation about this elsewhere... but I'll comment here too.

In general, it is safer if objects like URL() were used to construct the URLs instead of constructing them with string interpolation.
- The URL object handles a number of cases which are hard to do correctly. One is that query params can be repeated in a URL but using a JS dict doesn't allow duplicates. Another place it can help is safely joining paths together. Inserting values from untrusted input is safer than just using string interpolation (here we already use the searchParams though for the query params).
- the type checker when using an API vs using strings.
- The code is easier to review/audit.

Probably we have other parts in the codebase already that use strings for constructing URLs, so it could be done as a different PR or put in an issue I guess.

Reply:
What I meant in https://github.com/kinvolk/headlamp/pull/777#discussion_r1012785684 is that we cannot use the `URL` class here because we're still dealing with a partial URL (or rather its path + query), so creating an instance of `URL` with the point of validating it here would fail (the actual full URL is created later, when performing the actual request).
And since we already use `URLSearchParams` in `asQuery`, I don't think this logic is very unsafe as it is, or not more than before anyway.
So any further improvements to safety should be done at the request level and can be done in a separate PR.